### PR TITLE
Refactor(networkd): Replace networkd with a standalone app

### PR DESCRIPTION
**Pull request by bradbeam**
Signed-off-by: Brad Beam <brad.beam@talos-systems.com>

### CODE REVIEW

File: internal/pkg/kernel/kernel.go [hunk 1]
```diff
@@ -24,6 +24,8 @@ func NewDefaultCmdline() *Cmdline {
 	// See https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nvme-ebs-volumes.html.
 	cmdline.Append("nvme_core.io_timeout", "4294967295")
 	cmdline.Append("random.trust_cpu", "on")
+	// Disable rate limited printk
+	cmdline.Append("printk.devkmsg", "on")
```
Comment (bradbeam) [inline internal/pkg/kernel/kernel.go]: make this a separate pr

File: pkg/userdata/download.go [hunk 2]
```diff
@@ -8,6 +8,7 @@ import (
 	"encoding/base64"
 	"fmt"
 	"io/ioutil"
+	"log"
```
Comment (bradbeam) [inline pkg/userdata/download.go]: make this a separate pr

File: internal/pkg/kernel/kernel.go (hunk 1 above)
Comment (bradbeam) [reply internal/pkg/kernel/kernel.go]: #1027

File: pkg/userdata/download.go (hunk 2 above)
Comment (bradbeam) [reply pkg/userdata/download.go]: #1030

File: internal/app/networkd/main.go [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	)
+
+	nwd, err := networkd.New()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	// Is there any value in returning the low level link or
+	// should the translation to nic be handled here?
+	// links := nwd.Discover()
```
Comment (andrewrynhard) [inline internal/app/networkd/main.go]: Let's remove the commented code.

File: internal/app/networkd/pkg/address/address.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			dstLength, _ := route.Dest.Mask.Size()
+			routeMsg.DstLength = uint8(dstLength)
+		}
+
+		if route.Router != nil {
+			attr.Gateway = route.Router
+		}
+
+		// TODO figure out if Src is actually needed
+		/*
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/address/address.go]: Remove commented code.

File: internal/app/networkd/pkg/address/dhcp.go [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+		if *p != "azure" {
+			opts = append(opts, dhcpv4.OptionInterfaceMTU)
+		}
+	}
+
+	mods := []dhcpv4.Modifier{dhcpv4.WithRequestedOptions(opts...)}
+
+	// TODO expose this with some debug logging option
+	cli, err := nclient4.New(name, nclient4.WithTimeout(10*time.Second), nclient4.WithDebugLogger())
+	//cli, err := nclient4.New(name, nclient4.WithTimeout(2*time.Second))
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/address/dhcp.go]: Left over from testing?

File: internal/app/networkd/pkg/networkd/misc.go [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+// we need to
+func filterInterfaceByName(links []rtnetlink.LinkMessage) (filteredLinks []rtnetlink.LinkMessage) {
+	for _, link := range links {
+		switch {
+		case strings.HasPrefix(link.Attributes.Name, "en"):
+			filteredLinks = append(filteredLinks, link)
+		case strings.HasPrefix(link.Attributes.Name, "eth"):
+			filteredLinks = append(filteredLinks, link)
+			// TODO Add bond support
+			// case strings.HasPrefix(netif.Name, "bond"):
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/misc.go]: Remove commented code.

File: internal/app/networkd/pkg/networkd/netlink.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+				continue
+			}
+		}
+
+		// TODO
+		// Unsure what to do if scope doesnt match
+		// feels like it should require a delete + add
+
+		// TODO figure out if Src is actually needed
+		/*
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/netlink.go]: Remove commented code.

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 8]
```diff
@@ -0,0 +1,229 @@
...
+			wg.Add(1)
+
+			go n.renew(method, iface.Name, iface.Index)
+		}
+	}
+
+	// We dont ever wg.Done
+	// because this should run forever
+	// Probably a better way to do this
+	wg.Wait()
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/networkd.go]: There is ... an empty `select{}` is the better way here.

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 9]
```diff
@@ -0,0 +1,229 @@
...
+			if method.Hostname() != "" {
+				return method.Hostname()
+			}
+		}
+	}
+
+	return ""
+}
+
+/*
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/networkd.go]: I'd normally say remove commented code, but this one in particular looks like we can turn into some API data?

File: internal/app/networkd/pkg/nic/nic.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+import (
+	"errors"
+
+	"github.com/hashicorp/go-multierror"
+	"github.com/talos-systems/talos/internal/app/networkd/pkg/address"
+)
+
+// TODO: Probably should put together some map here
+// to map number with string()
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/nic/nic.go]: Probably a good idea. Here is the pattern used throughout the rest of our code:

```
// Type represents an interface type.
type Type int

const (
	Single Type = iota
	Bond
)

// String returns the string representation of a Type.
func (t Type) String() string {
	return [...]string{"Single", "Bond"}[t]
}
```

File: internal/app/networkd/pkg/nic/nic.go [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+	Bond = iota
+	Single
+
+	// https://tools.ietf.org/html/rfc791
+	MinimumMTU = 68
+	MaximumMTU = 65536
+)
+
+// NetworkInterface provides an abstract configuration representation for a
+// network interface
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/nic/nic.go]: As a general rule of thumb, Public funcs/methods require a comment, and we should comment with proper grammar (ending in a period in this case) since this ends up the public GoDoc site. More of a note for the future.

File: internal/app/networkd/pkg/nic/nic.go [hunk 12]
```diff
@@ -0,0 +1,63 @@
...
+	MTU           uint32
+	Index         uint32
+	SubInterfaces []string
+	AddressMethod []address.Addressing
+}
+
+// Create returns a NetworkInterface with all of the given setter options
+// applied
+func Create(setters ...Option) (*NetworkInterface, error) {
+
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/nic/nic.go]: Extra line.

File: internal/app/networkd/pkg/nic/options.go [hunk 13]
```diff
@@ -0,0 +1,85 @@
...
+			n.Type = Single
+		default:
+			return errors.New("unsupported network interface type")
+		}
+		return err
+	}
+}
+
+// WithMTU defines the MTU for the interface
+// TODO: I think we should drop this since MTU is getting set
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/nic/options.go]: It _may_ be useful in your API.

File: internal/app/machined/internal/phase/network/network.go [hunk 14]
```diff
@@ -33,25 +36,58 @@ func (task *UserDefinedNetwork) RuntimeFunc(mode runtime.Mode) phase.RuntimeFunc
...
 	var hostname string
 	kernelHostname := kernel.ProcCmdline().Get(constants.KernelParamHostname).First()
 	switch {
 	case data.Networking != nil && data.Networking.OS != nil && data.Networking.OS.Hostname != "":
 		hostname = data.Networking.OS.Hostname
 	case kernelHostname != nil:
 		hostname = *kernelHostname
+	case nwd.Hostname() != "":
+		hostname = nwd.Hostname(netIfaces...)
 	default:
```
Comment (andrewrynhard) [inline internal/app/machined/internal/phase/network/network.go]: We can skip the empty `default`.

File: internal/app/machined/pkg/system/services/networkd.go [hunk 15]
```diff
@@ -19,31 +25,62 @@ import (
...
+
+	env := []string{}
+	for key, val := range data.Env {
+		env = append(env, fmt.Sprintf("%s=%s", key, val))
+	}
+
+	return restart.New(containerd.NewRunner(
+		data,
+		&args,
+		runner.WithContainerImage(image),
```
Comment (andrewrynhard) [inline internal/app/machined/pkg/system/services/networkd.go]: Now that system containerd was merged, we should run this using that instance.

File: internal/app/machined/pkg/system/services/networkd.go [hunk 16]
```diff
@@ -19,31 +25,62 @@ import (
...
-func (c *Networkd) Condition(data *userdata.UserData) conditions.Condition {
+func (n *Networkd) Condition(data *userdata.UserData) conditions.Condition {
 	return nil
 }
 
 // DependsOn implements the Service interface.
-func (c *Networkd) DependsOn(data *userdata.UserData) []string {
-	return nil
+func (n *Networkd) DependsOn(data *userdata.UserData) []string {
+	return []string{"containerd"}
```
Comment (andrewrynhard) [inline internal/app/machined/pkg/system/services/networkd.go]: system-containerd

File: internal/app/networkd/main.go [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+	// kick off the addressing mechanism
+	// Add any necessary routes
+	log.Println("configuring interface addressing")
+	if err = nwd.Configure(netIfaces...); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println("starting renewal watcher")
+	// handle dhcp renewal
+	nwd.Renew(netIfaces...)
```
Comment (andrewrynhard) [inline internal/app/networkd/main.go]: Will the gRPC server be in a follow up PR?

Comment (bradbeam) [reply internal/app/networkd/main.go]: yeah

File: internal/app/networkd/pkg/networkd/netconf.go [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+			}
+
+		}
+	}
+
+	return nil
+}
+
+// validateNetworkUserData ensures that we have actual data to do our lookups
+func validNetworkUserData(data *userdata.UserData) bool {
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/netconf.go]: `s/validNetworkUserData/validateNetworkUserData

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 19]
```diff
@@ -0,0 +1,274 @@
...
+)
+
+// Set up default nameservers
+const (
+	DefaultPrimaryResolver   = "1.1.1.1"
+	DefaultSecondaryResolver = "8.8.8.8"
+)
+
+// Networkd provides the high level interaction to configure network interfaces
+// on a host system. This currently support addressing configuration via dhcp
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/networkd.go]: `s/This currently support/This currently supports`

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 20]
```diff
@@ -0,0 +1,274 @@
...
+// the address discovery ( static vs dhcp ) as well as the netlink interaction
+// to set an address on the link and create any routes.
+func (n *Networkd) Configure(ifaces ...*nic.NetworkInterface) error {
+	var (
+		err       error
+		resolvers []net.IP
+	)
+
+	for _, iface := range ifaces {
+		//log.Printf("configuring %+v\n", iface)
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/networkd.go]: Remove

File: internal/app/networkd/pkg/networkd/networkd.go [hunk 21]
```diff
@@ -0,0 +1,274 @@
...
+	}
+
+	return nil
+}
+
+// Renew sets up a long running loop to refresh a network interfaces
+// addressing configuration. Currently this only applies to interfaces
+// configured by DHCP.
+func (n *Networkd) Renew(ifaces ...*nic.NetworkInterface) {
+	var wg sync.WaitGroup
```
Comment (andrewrynhard) [inline internal/app/networkd/pkg/networkd/networkd.go]: We can remove the WaitGroup since we don't use it for waiting.

File: internal/app/networkd/main.go [hunk 22]
```diff
@@ -0,0 +1,75 @@
...
+
+	// kick off the addressing mechanism
+	// Add any necessary routes
+	log.Println("configuring interface addressing")
+	if err = nwd.Configure(netIfaces...); err != nil {
+		log.Fatal(err)
+	}
+
+	log.Println("interface configuration")
+	nwd.PrintState()
```
Comment (andrewrynhard) [inline internal/app/networkd/main.go]: In general I'm not a big fan of printing out raw data structures, but I think this is fine for now since I imagine we will remove this and turn it into an API call.

Comment (bradbeam) [reply internal/app/networkd/main.go]: yeah, until we get the api built up, this is the quickest win to show the current state
